### ISSUE TITLE
Question about GRPO-guard implementation details (rationorm behavior)

### ISSUE THREAD

**Issue by simonat2011** [yifan123/flow_grpo#223]
Hi, thanks for open-sourcing this repo!

I have a question regarding the implementation details of **GRPO-guard**, specifically about the `rationorm` part.

In the current implementation, when `config.rationorm` is enabled, it seems that `ratio_mean_bias` and `sigma_t` (as well as `sqrt_dt`) are **averaged first**, and then this averaged term is added into the ratio computation. Concretely, `ratio_mean_bias` is reduced by `.mean(...)`, `sigma_t = std_dev_t.mean()`, and `sqrt_dt.mean()` is also used before forming the final ratio.

However, from my reading of the paper, this behavior is not very explicitly described. The paper seems to suggest that the bias term is computed at the per-sample level first, and then the ratio is averaged afterward, rather than averaging the statistics (`sigma_t`, `sqrt_dt`, and the bias) before applying them to the ratio.

So I’m wondering whether there is a specific design or stability reason for averaging `ratio_mean_bias`, `sigma_t`, and `sqrt_dt` **before** incorporating them into the ratio?

For reference, here is the relevant code snippet in `train_sd3_GRPO_Guard.py`:

```python
if config.rationorm:
    sigma_t = std_dev_t.mean()
    ratio_mean_bias = (
        (prev_sample_mean - sample["prev_sample_mean"][:, j])
        .pow(2)
        .mean(dim=tuple(range(1, log_prob.ndim)))
    )
    ratio_mean_bias = ratio_mean_bias / (2 * (sqrt_dt.mean() * sigma_t) ** 2)
    ratio = torch.exp(
        (log_prob - sample["log_probs"][:, j] + ratio_mean_bias)
        * (sqrt_dt.mean() * sigma_t)
    )
else:
    ratio = torch.exp(log_prob - sample["log_probs"][:, j])
```

Any clarification on the rationale behind this implementation choice would be greatly appreciated. Thanks!

**Comment by jingw193** [comment]
Thank you for your interest in GRPO-Guard. In practice, `std_dev_t` and `sqrt_dt` depend only on the diffusion scheduler and are identical across all samples. Therefore, for computational convenience, we directly take their mean for subsequent calculations.
In contrast, the sample-specific component is `ratio_mean_bias`, for which the batch dimension is preserved when computing the mean (`.mean(dim=tuple(range(1, log_prob.ndim)))`).

**Comment by simonat2011** [comment]
Thanks a lot for the clarification! Really appreciate the detailed explanation and the quick response!
